src/nupic/algorithms/Connections.cpp

```diff
@@ -206,11 +203,13 @@ void Connections::destroySynapse(Synapse synapse) {
 
 void Connections::updateSynapsePermanence(Synapse synapse,
                                           Permanence permanence) {
-  for (auto h : eventHandlers_) {
-    h.second->onUpdateSynapsePermanence(synapse, permanence);
-  }
-
+  permanence = std::min(permanence, maxPermanence );
```

Comment:
I think this is better design for function (`updateSynapsePermanence`) to clip the permanence into the correct range, and no one else should clip permanences.  I'm pretty sure that every piece of code which modifies permanences needs to apply this clipping to the result.

Reply:
my idea was microoptimization - to avoid 1op for max, 1op for assign. I wanted to validate that in Conn Permanence is always within the [0,1] range, so even this clipping could be removed. 
(the assert would be there to make code clearer, check in debug)
But maybe that's overcomplicating things

Reply:
I think I understand what your are talking about.  I think that you always need to check the permanence when it gets modified, but that partial information can help.  Multiple variants of this function could be useful and optimized:
* We could have an increment variant, and a decrement variant, and these would be able to omit a single bounds check each.  
* A variant with both checks for creature comfort.
* A common version with no checks, for the other variants to call.

Reply:
...let's get this PR to merge and then see again in #161